### DISCUSSION TITLE
1.5.1-beta6

### DISCUSSION THREAD

**Discussion by ilyhalight** [ilyhalight/voice-over-translation#526] [Обновления]
## Список изменений

- Теперь, полученные переводы кешируются до перезагрузки страницы. Благодаря этому удалось избавиться от лишних запросов к серверу, а так же это избавило от проблем, когда из-за изменения качества приходилось заново ждать перевод (например на weverse)
- Старая реализация установки громкости звука в "stopTranslate" изменена на актуальную - "this.setVideoVolume"
- Объединен некоторый повторяющийся код

**Full Changelog**: https://github.com/ilyhalight/voice-over-translation/compare/1.5.1-beta5...1.5.1-beta6

<hr /><em>This discussion was created from the release <a href='https://github.com/ilyhalight/voice-over-translation/releases/tag/1.5.1-beta6'>1.5.1-beta6</a>.</em>

**Comment by NullVerdict** [comment]
Недавно я заметил, что перевод трансляций иногда прерывается на буферизацию.  Это обычно происходит, когда количество зрителей, использующих перевод, превышает одного человека.

**Comment by ilya-corneli** [comment]
Проблемная версия, стало переводить с английского на английский.
Выдает ошибку иногда и тупо не переводит

**Reply by ilyhalight** [reply]
При каких условиях появляются проблемы?

**Comment by NullVerdict** [comment]
@ilyhalight Яндекс обновлял список языков?

**Reply by FiLinX** [reply]
чего то у меня не получается из яндекс браузера (https://download.cdn.yandex.net/browser/switch-brand/browser-setup.arc) в вытащенной папке video_translation просмотреть файлы..
они закодированы как то чтоли?
вот например как выклядит открытый в блокноте файл оттуда supported_languages_by_feature  -  https://i.imgur.com/KriyONx.png
сплошь кракозябры, будто это бинарный просматриваю, но при открытии нету предупреждения о бинарности..

**Reply by ilyhalight** [reply]
> чего то у меня не получается из яндекс браузера (https://download.cdn.yandex.net/browser/switch-brand/browser-setup.arc) в вытащенной папке video_translation просмотреть файлы..
> они закодированы как то чтоли?
> вот например как выклядит открытый в блокноте файл оттуда supported_languages_by_feature  -  https://i.imgur.com/KriyONx.png
> сплошь кракозябры, будто это бинарный просматриваю, но при открытии нету предупреждения о бинарности..

не помню, то ли это их собственный какой-то формат ytxt, либо из-за сборки в .arc так происходит

**Reply by ilyhalight** [reply]
> @ilyhalight Яндекс обновлял список языков?

Вроде бы нет, сейчас не могу проверить

**Reply by NullVerdict** [reply]
Кстати, может быть в будущем попробывать использовать [webpack-monkey](https://github.com/guansss/webpack-monkey)?

**Reply by ilyhalight** [reply]
> Кстати, может быть в будущем попробывать использовать [webpack-monkey](https://github.com/guansss/webpack-monkey)?

Переписал на него, как минимум дев режим там удобнее сделан

**Reply by NullVerdict** [reply]
> > Кстати, может быть в будущем попробывать использовать [webpack-monkey](https://github.com/guansss/webpack-monkey)?
> 
> Переписал на него, как минимум дев режим там удобнее сделан

Скрипт перестал работать, ошибка `__MK_GLOBAL__` is not defined

**Reply by NullVerdict** [reply]
Разобрался, пока работает только dev

**Reply by ilyhalight** [reply]
> 
> Скрипт перестал работать, ошибка `__MK_GLOBAL__` is not defined

Исправил, в плагине было задумано, что `__MK_GLOBAL__` будет задаваться для `window.__MK_GLOBAL__`, но почему-то это не работает. Пришлось через вебпак вручную задать

**Reply by NullVerdict** [reply]
Кстати, grant начали дублироваться, но это не критично

**Reply by NullVerdict** [reply]
Я заметил еще одну вещь: yandexProtobuf плохо работает с CSP.
При вызове yandexProtobuf.encodeTranslationRequest выводится CSP-ошибка. Возможно, проблема в Protobuf.js.
Может быть, стоит добавить расчет Protobuf в сам прокси? Возможно, это поможет решить CSP-ошибку.
Также в будующем можно заменить все fetch на какой нибудь GM-Fetch или на обычный GM_xmlhttpRequest, дабы избавиться от всех CSP ошибок. 
```js
    const body = yandexProtobuf.encodeTranslationRequest(
      url,
      duration,
      requestLang,
      responseLang,
      translationHelp,
    );
```

**Reply by ilyhalight** [reply]
> Может быть, стоит добавить расчет Protobuf в сам прокси

Тоже думал насчет переноса логики на сам прокси, надо будет в будущем попробовать так сделать

> Кстати, grant начали дублироваться, но это не критично

Оказывается плагин автоматически подтягивает все нужные гранты из кода, но не проверяет, что они уже указаны в мете и из-за этого появлялись дубли. Исправил

**Comment by ilya-corneli** [comment]
Не удалось запросить перевод видео, перезагрузка не помогла, не работает на всех видео, проблема началась сегодня

**Reply by ilyhalight** [reply]
> Не удалось запросить перевод видео, перезагрузка не помогла, не работает на всех видео, проблема началась сегодня

cloudflare версия? Если да, то поменяйте в настройках расширения домен на один из доступных. На рендере лимит достигнут

![Screenshot_2024-02-25-11-11-27-27_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/ilyhalight/voice-over-translation/assets/62353659/20b53485-9314-4de8-975f-eff4b7729fb2)

**Reply by ilyhalight** [reply]
> > Не удалось запросить перевод видео, перезагрузка не помогла, не работает на всех видео, проблема началась сегодня
> 
> cloudflare версия? Если да, то поменяйте в настройках расширения домен на один из доступных. На рендере лимит достигнут
> 
> ![Screenshot_2024-02-25-11-11-27-27_984e1414ae90666a90f12ff17ec14a7f.jpg](https://github.com/ilyhalight/voice-over-translation/assets/62353659/20b53485-9314-4de8-975f-eff4b7729fb2)
> 
> 

сервер перевода тоже что-то упал, надо будет разобраться

**Reply by NullVerdict** [reply]
Надо сделать 3 сервера и рандомно между ними переключаться. Ну или поставить лимиты для спамеров, чтобы не перегружали.

**Reply by ilyhalight** [reply]
Наверное, сделаю через какой-нибудь нгинксовский лоад балансер, чтобы если 1 из серверов был недоступен, то запрос не уходил бы в пустоту, как раз и проблему с региональными ограничениями воркеров решит

> поставить лимиты для спамеров

Мне кажется тут скорее не спамеры виноваты, а проксирование аудио, поскольку в случае рендера забился именно лимит трафика в 100 гигов

**Reply by NullVerdict** [reply]
> Наверное, сделаю через какой-нибудь нгинксовский лоад балансер, чтобы если 1 из серверов был недоступен, то запрос не уходил бы в пустоту, как раз и проблему с региональными ограничениями воркеров решит
> 
> > поставить лимиты для спамеров
> 
> Мне кажется тут скорее не спамеры виноваты, а проксирование аудио, поскольку в случае рендера забился именно лимит трафика в 100 гигов

В этом случае нужно указать другой домен, который не зависит от трафика. Лимит в 100GB явно не предназначен для файлов с размером более 10MB, лимит в лучшем случае будет 10к проксированных аудио. Тут скорее просчёт. На Deno оставить проксирование аудио а Render оставить как основной прокси Яндекса.

**Reply by ilyhalight** [reply]
> > Наверное, сделаю через какой-нибудь нгинксовский лоад балансер, чтобы если 1 из серверов был недоступен, то запрос не уходил бы в пустоту, как раз и проблему с региональными ограничениями воркеров решит
> > > поставить лимиты для спамеров
> > 
> > 
> > Мне кажется тут скорее не спамеры виноваты, а проксирование аудио, поскольку в случае рендера забился именно лимит трафика в 100 гигов
> 
> В этом случае нужно указать другой домен, который не зависит от трафика. Лимит в 100GB явно не предназначен для файлов с размером более 10MB, лимит в лучшем случае будет 10к проксированных аудио. Тут скорее просчёт. На Deno оставить проксирование аудио а Render оставить как основной прокси Яндекса.

Пока сделал просто общий балансер между всеми доступными доменами, чтобы запросы не так быстро забивались.

> На Deno оставить проксирование аудио а Render оставить как основной прокси Яндекса.

Там тоже есть лимит в 100гб, поэтому особого толку не будет

**Comment by Bioswamp** [comment]
Сори...а в чем отличие клауд версии vot-cloudflare.user.js от vot-cloudflare-min.user.js ? чтото ненайду где написано

**Reply by ilyhalight** [reply]
min - минифицированная версия, она весит меньше, но ее код не читабельный для человека

**Comment by Wintreist** [comment]
Возможно ли как-то сделать так, чтобы можно было переводить видео с локального компьютера?
Допустим я скачал что-то с авторскими правами с торрента и поэтому не могу загрузить на ютуб, чтобы его перевело

**Reply by ilyhalight** [reply]
яндекс диск, гугл диск и прочее в помощь. С локального никак

**Reply by Wintreist** [reply]
> яндекс диск, гугл диск и прочее в помощь. С локального никак

И можно потом зайти на само видео и перевести его? Нужны какие-нибудь настройки конфиденциальности? Что-то вроде доступа по ссылке?

**Reply by ilyhalight** [reply]
да, нужно сделать доступ по ссылке. После нужно перейти по этой ссылке и уже оттуда переводить

**Comment by ilya-corneli** [comment]
www.banned.video
Не переводит, не реагирует на нажатие кнопки

**Reply by ilyhalight** [reply]
Все работает, проверьте, что используете актуальную версию скрипта. Если у вас cloudflare версия, то поставьте vot.toil.cc в поле с "адресом worker прокси-сервера" в меню расширения

**Reply by ilya-corneli** [reply]
спасибо заработало после переустановки, извините за беспокойство
